services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java

```diff
@@ -0,0 +1,281 @@
+                      storeName,
+                      targetVersionNum,
+                      targetRegion);
+                }
+              }
+
+              if (!didPushCompleteInTargetRegions) {
+                continue;
+              }
+
+              // Check that push is complete in non target regions
+              int numNonTargetRegionsCompleted = remainingRegions.size();
+              for (String remainingRegion: remainingRegions) {
+                String executionStatus = pushStatusInfo.getExtraInfo().get(remainingRegion);
+                if (!executionStatus.equals(ExecutionStatus.COMPLETED.toString())) {
+                  numNonTargetRegionsCompleted -= 1;
+                  LOGGER.info(
+                      "Push is not complete for store: {} on version: {} in a non target region: {}",
+                      storeName,
+                      targetVersionNum,
+                      remainingRegion);
+                }
+              }
+
+              // If the majority of the remaining regions have failed their push jobs, mark the version status as ERROR
+              // so that we don't check this store again for this version
+              HelixVeniceClusterResources resources =
+                  veniceParentHelixAdmin.getVeniceHelixAdmin().getHelixVeniceClusterResources(cluster);
+              ReadWriteStoreRepository repository = resources.getStoreMetadataRepository();
+              Store parentStore = repository.getStore(storeName);
+              if (numNonTargetRegionsCompleted < remainingRegions.size() / 2) {
+                LOGGER.warn(
+                    "Skipping version swap for store: {} on version: {} as majority of non target regions are not complete",
+                    storeName,
+                    targetVersionNum);
+                parentStore.updateVersionStatus(targetVersionNum, ERROR);
+                repository.updateStore(parentStore);
+                continue;
+              }
+
+              // Check that waitTime has elapsed in target regions
+              boolean didWaitTimeElapseInTargetRegions = didWaitTimeElapseInTargetRegions(
+                  pushStatusInfo.getExtraInfoUpdateTimestamp(),
+                  targetRegions,
+                  store.getTargetSwapRegionWaitTime());
+
+              if (!didWaitTimeElapseInTargetRegions) {
+                LOGGER.info(
+                    "Skipping version swap for store: {} on version: {} as wait time: {} has not passed",
+                    storeName,
+                    targetVersionNum,
+                    store.getTargetSwapRegionWaitTime());
+                storePushCompletionTimeCache.put(storeName, pushStatusInfo.getExtraInfoUpdateTimestamp());
+                continue;
+              }
+
+              // TODO add call for postStoreVersionSwap() once it is implemented
+
+              String remainingRegionsString = String.join(",\\s*", remainingRegions);
```

Comment:
We need to filter out failed regions from the roll-forward command, we do not want to switch the version in those regions.